controlplane/pkg/prefix_pool/prefixpool.go

```diff
@@ -47,6 +51,121 @@ func NewPrefixPool(prefixes ...string) (PrefixPool, error) {
 	}, nil
 }
 
+/* Release excluded prefixes back the pool of available ones */
+func (impl *prefixPool) ReleaseExcludedPrefixes(excludedPrefixes []string) error {
+	impl.Lock()
+	defer impl.Unlock()
+
+	remaining, err := ReleasePrefixes(impl.prefixes, excludedPrefixes...)
+	if err != nil {
+		return err
+	}
+	impl.prefixes = remaining
+	return nil
+}
+
+/* Exclude prefixes from the pool of available prefixes */
+func (impl *prefixPool) ExcludePrefixes(excludedPrefixes []string) ([]string, error) {
+	impl.Lock()
+	defer impl.Unlock()
+	/* Use a working copy for the available prefixes */
+	copyPrefixes := append([]string{}, impl.prefixes...)
+
+	removedPrefixes := []string{}
+
+	for _, excludedPrefix := range excludedPrefixes {
+		splittedEntries := []string{}
+		prefixesToRemove := []string{}
+		_, subnetExclude, _ := net.ParseCIDR(excludedPrefix)
+
+		/* 1. Check if each excluded entry overlaps with the available prefix */
+		for _, prefix := range copyPrefixes {
+			intersecting := false
+			excludedIsBigger := false
+			_, subnetPrefix, _ := net.ParseCIDR(prefix)
+			intersecting, excludedIsBigger = intersect(subnetExclude, subnetPrefix)
+			/* 1.1. If intersecting, check which one is bigger */
+			if intersecting {
+				/* 1.1.1. If excluded is bigger, we remove the original entry */
```

Comment:
@rdimitrov Hm, probably I am mistaken, but if we simply remove original entry in that case, then during releasing we increase the original size of prefix pool. I mean the following test most likely will fail:
```go
pool, err := NewPrefixPool("10.20.0.0/16", "2.20.0.0/16")
Expect(err).To(BeNil())
excludedPrefix := []string{"2.20.0.0/8"}

excluded, err := pool.ExcludePrefixes(excludedPrefix)

Expect(err).To(BeNil())
Expect(pool.GetPrefixes()).To(Equal([]string{"10.20.0.0/16"}))

err = pool.ReleaseExcludedPrefixes(excluded)
Expect(err).To(BeNil())
Expect(pool.GetPrefixes()).To(Equal([]string{"10.20.0.0/16", "2.20.0.0/16"}))
```

Reply:
Thanks, @lobkovilya! 👍 I've updated the tests with that use case and added a fix for it.

Reply:
By the way, I found this one to be really useful for verifying what to expect - https://www.techzoom.net/Tools/IPAddressCalculator 👍